Firestore/core/src/local/leveldb_document_overlay_cache.cc

```diff
@@ -258,6 +250,40 @@ void LevelDbDocumentOverlayCache::ForEachKeyWithLargestBatchId(
   }
 }
 
+void LevelDbDocumentOverlayCache::ForEachKeyInCollection(
+    const ResourcePath& collection,
+    int since_batch_id,
+    std::function<void(LevelDbDocumentOverlayKey&&)> callback) const {
+  const std::string index_start_key =
+      LevelDbDocumentOverlayCollectionIndexKey::KeyPrefix(user_id_, collection,
+                                                          since_batch_id + 1);
+  const std::string index_key_prefix =
+      LevelDbDocumentOverlayCollectionIndexKey::KeyPrefix(user_id_, collection);
+
+  auto it = db_->current_transaction()->NewIterator();
+  for (it->Seek(index_start_key);
+       it->Valid() && absl::StartsWith(it->key(), index_key_prefix);
+       it->Next()) {
+    LevelDbDocumentOverlayCollectionIndexKey key;
+    HARD_ASSERT(key.Decode(it->key()));
+    if (key.collection() != collection) {
```

Comment:
Will this work with nested subcollections that share the same prefix? We probably need to "continue" here like in https://osscs.corp.google.com/firebase-sdk/firebase-ios-sdk/+/master:Firestore/core/src/local/leveldb_mutation_queue.cc;l=323

We already validate that this loop does not go "out of bounds" with the check in line 265.

Reply:
I believe it will work in that case because in the encoding of the key the collection is followed by the largest_batch_id. What would be a test case that I could write? I'm not sure I 100% understand the scenario you're describing.

I also want to avoid, at all costs, using `continue` instead of `break` because if using `continue` then the performance of this method degrades to a full table scan in the worst case.

Reply:
After discussing this offline I think it should work. The collection path is written in segments, and the segments are terminated by a character that cannot be part of a segment. So the key prefix you are using will always end in a terminator for the last path segment, then the component label for the batch ID and then the batch ID. The only cause of conflict would be if the the first byte after the terminator were the batch ID label, but that is not possible since it would have to be the component label for a path component, which will never match the batch ID label.

That being said, I still don't think you need this if check. Your loop should terminate based on `absl::StartsWith(it->key(), index_key_prefix)`.

Reply:
The `if` check is still required because `index_key_prefix` does _not_ include the largest_batch_id and therefore it is also a prefix of the keys for overlays subcollections.

If I comment out that "if" check then the `GetAllOverlaysForCollection` test fails at line 255 because the call to `GetOverlays()` returns the overlays for the requested collection AND its subcollections (it should NOT return overlays for the subcollections, only the collection itself).

Reply:
Okay, let's keep it as is. The alternative of adding an additional marker or including the marker of the next component in the key prefix is not very appealing. Thanks for thinking this through with me.